File: service/app/src/test/java/io/miragon/blueprint/application/service/SendContractServiceTest.java
package io.miragon.blueprint.application.service;

import static io.miragon.blueprint.domain.leasing.TestObjectBuilder.testLeasingApplication;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.ArgumentMatchers.argThat;
import static org.mockito.ArgumentMatchers.eq;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.verifyNoMoreInteractions;
import static org.mockito.Mockito.when;

import io.miragon.blueprint.application.port.outbound.ContractPort;
import io.miragon.blueprint.application.port.outbound.LeasingApplicationRepository;
import io.miragon.blueprint.application.port.outbound.NotificationPort;
import io.miragon.blueprint.domain.leasing.ContractId;
import io.miragon.blueprint.domain.leasing.LeasingApplication;
import java.util.Optional;
import org.junit.jupiter.api.Test;

class SendContractServiceTest {

    private final LeasingApplicationRepository repository = mock(LeasingApplicationRepository.class);
    private final ContractPort contract = mock(ContractPort.class);
    private final NotificationPort notification = mock(NotificationPort.class);
    private final SendContractService underTest =
            new SendContractService(repository, contract, notification);

    @Test
    void sendContractIssuesTheContractRecordsItsIdOnTheApplicationAndNotifiesTheCustomer() {

        // given: an application whose contract the contract system will issue
        LeasingApplication application = testLeasingApplication().build();
        when(repository.findById(application.id())).thenReturn(Optional.of(application));
        when(contract.issueContract(application.id())).thenReturn(new ContractId("CONTRACT-1"));
        when(repository.save(any())).thenAnswer(invocation -> invocation.getArgument(0));

        // when: the contract is sent
        underTest.sendContract(application.id());

        // then: the contract is issued, its id is stored on the application and the customer is asked to sign
        verify(repository).findById(application.id());
        verify(contract).issueContract(application.id());
        verify(repository).save(argThat(app -> new ContractId("CONTRACT-1").equals(app.contractId())));
        verify(notification).send(any(), eq(application));
        verifyNoMoreInteractions(repository, contract, notification);
    }
}

File: service/app/src/test/kotlin/io/miragon/blueprint/application/service/SendContractServiceTest.kt
package io.miragon.blueprint.application.service

import io.miragon.blueprint.application.port.outbound.ContractPort
import io.miragon.blueprint.application.port.outbound.LeasingApplicationRepository
import io.miragon.blueprint.application.port.outbound.NotificationPort
import io.miragon.blueprint.domain.leasing.ContractId
import io.miragon.blueprint.domain.leasing.testLeasingApplication
import io.mockk.Runs
import io.mockk.confirmVerified
import io.mockk.every
import io.mockk.just
import io.mockk.mockk
import io.mockk.verify
import org.junit.jupiter.api.Test

class SendContractServiceTest {

    private val repository = mockk<LeasingApplicationRepository>()
    private val contract = mockk<ContractPort>()
    private val notification = mockk<NotificationPort>()
    private val underTest =
        SendContractService(repository = repository, contract = contract, notification = notification)

    @Test
    fun `sendContract issues the contract, records its id on the application and notifies the customer`() {

        // given: an application whose contract the contract system will issue
        val application = testLeasingApplication()
        every { repository.findById(application.id) } returns application
        every { contract.issueContract(application.id) } returns ContractId("CONTRACT-1")
        every { repository.save(any()) } answers { firstArg() }
        every { notification.send(any(), any()) } just Runs

        // when: the contract is sent
        underTest.sendContract(application.id)

        // then: the contract is issued, its id is stored on the application and the customer is asked to sign
        verify { repository.findById(application.id) }
        verify { contract.issueContract(application.id) }
        verify { repository.save(match { it.contractId == ContractId("CONTRACT-1") }) }
        verify { notification.send(any(), application) }
        confirmVerified(repository, contract, notification)
    }
}
